src/web/arr/trove/chart.arr

```diff
@@ -0,0 +1,1028 @@
+
+  default-chart-window-object.{
+    method render(self) block:
+      is-infer-x = unsafe-equal(self.x-min, 0) and unsafe-equal(self.x-max, 0)
+      is-infer-y = unsafe-equal(self.y-min, 0) and unsafe-equal(self.y-max, 0)
+
+      combined-pts = if is-infer-x or is-infer-y:
+        (line-plots.map(_.ps) + scatter-plots.map(_.ps)).foldl(_ + _, empty)
+      else:
+        empty
+      end
+
+      shadow combined-pts = if not(is-infer-x):
+        combined-pts.filter(in-bound-x(_, self))
+      else:
+        combined-pts
+      end
+
+      shadow combined-pts = if not(is-infer-y):
+        combined-pts.filter(in-bound-y(_, self))
+      else:
+        combined-pts
+      end
+
+      maybe-bbox = if is-infer-x or is-infer-y:
+        some(get-bounding-box(combined-pts))
+      else:
+        none
+      end
+
+      shadow self = if is-infer-x:
+        cases (Option) maybe-bbox:
+          | none => raise('impossible')
+          | some(bbox) =>
+            {x-min; x-max} = if bbox.is-valid:
+              widen-range(bbox.x-min, bbox.x-max)
+            else:
+              DEFAULT-RANGE
+            end
+            self.{x-min: x-min, x-max: x-max}
+        end
+      else:
+        self
+      end
+
+      function-plots-data = function-plots
+        .map(generate-xy(_, self.x-min, self.x-max, self.num-samples))
+        .map(_.get-data())
+
+      shadow self = if is-infer-y:
+        cases (Option) maybe-bbox:
+          | none => raise('impossible')
+          | some(bbox) =>
+            function-plots-pts = function-plots-data.map(_.ps).foldl(_ + _, empty)
+            bbox2 = get-bounding-box(function-plots-pts)
+            {y-min; y-max} = if bbox.is-valid or bbox2.is-valid:
+              {y-min; y-max} = if bbox.is-valid and bbox2.is-valid:
+                {num-min(bbox.y-min, bbox2.y-min); num-max(bbox.y-max, bbox2.y-max)}
+                else if bbox.is-valid:
+                {bbox.y-min; bbox.y-max}
+                else:
+                {bbox2.y-min; bbox2.y-max}
+              end
+              widen-range(y-min, y-max)
+            else:
+              DEFAULT-RANGE
+            end
+            self.{y-min: y-min, y-max: y-max}
+        end
+      else:
+        self
+      end
+
+      fun helper(shadow self, shadow function-plots-data :: Option) -> IM.Image block:
+        shadow function-plots-data = cases (Option) function-plots-data:
+          | none => function-plots
+              .map(generate-xy(_, self.x-min, self.x-max, self.num-samples))
+              .map(_.get-data())
+          | some(shadow function-plots-data) => function-plots-data
+        end
+
+        scatters-arr = for map(p from scatter-plots + function-plots-data):
+          ps-to-arr(p.{ps: p.ps.filter(in-bound-xy(_, self))})
+        end ^ reverse ^ builtins.list-to-raw-array
+
+        # NOTE: lines-arr has JavaScript's null in it.
+        # Don't try to inspect it from Pyret's side
```

Comment:
WAT?  This sounds very very brittle to me...

Reply:
Yup, sorry. This is really just a hack for now (since @schanzer seems to want the library to finish soon). It will be gone eventually when I have time to rework this part.

Reply:
This really worries me; brittle hacks have an annoying tendency to stick around longer than we like.

Reply:
Agree with Ben. “There's nothing quite as permanent as a temporary fix”